src/test/java/guide/UsingHttpsTest.java

```diff
@@ -56,29 +68,133 @@ public void shouldPassWhenExpectedStubDidHappen() throws GeneralSecurityExceptio
         server.run();
+        whenHttp(server).match(get("/asd")).then(ok()).mustHappen(0);
+
+        try {
+            sslReadyHttpClient(null).execute(new HttpGet("https://localhost:" + server.getPort() + "/asd"));
+            fail("Request should have failed as client does not trust server's TLS certificate");
+        } catch (Exception e) {
+            assertThat(e, isA(SSLHandshakeException.class));
+        }
+
+        ensureHttp(server).gotStubsCommitmentsDone();
+    }
+
+    @Test
+    public void shouldFailWhenServerRequiresAuthentication() throws IOException {
+        server.clientAuth(Files.readAllBytes(Path.of("build/resources/test/keystore_server_test_cert")), "changeit").run();
+        whenHttp(server).match(get("/asd")).then(ok()).mustHappen(0);
+
+        try {
+            sslReadyHttpClient().execute(new HttpGet("https://localhost:" + server.getPort() + "/asd"));
+            fail("Request should have failed as server requires client authentication");
+        } catch (Exception e) {
+            assertThat(e, isA(IOException.class));
+        }
+
+        ensureHttp(server).gotStubsCommitmentsDone();
+    }
+
+    @Test
+    public void shouldPassWhenSendingCertificateToServerRequiringAuthentication() throws GeneralSecurityException, IOException {
+        // client authentication
+        server.clientAuth(Files.readAllBytes(Path.of("build/resources/test/keystore_server_test_cert")), "changeit").run();
         whenHttp(server).match(get("/asd")).then(ok()).mustHappen();
 
-        HttpResponse execute = sslReadyHttpClient().execute(new HttpGet("https://localhost:" + server.getPort() + "/asd"));
+        InputStream clientKeyStore = getClass().getResourceAsStream("/keystore_server_test");
+        String clientKeyStorePass = "secret";
+        HttpResponse execute = sslReadyHttpAuthClient(clientKeyStore, clientKeyStorePass).execute(new HttpGet("https://localhost:" + server.getPort() + "/asd"));
 
         assertThat(execute.getStatusLine().getStatusCode(), equalTo(200));
         ensureHttp(server).gotStubsCommitmentsDone();
     }
 
+    @Test
+    public void shouldFailSendingNonAllowedCertificate() throws IOException {
+        // client authentication
+        server.clientAuth(Files.readAllBytes(Path.of("build/resources/test/keystore_server_test_cert")), "changeit").run();
+        whenHttp(server).match(get("/asd")).then(ok()).mustHappen(0);
+
+        // sending a client certificate that is not allowed
+        InputStream clientKeyStore = StubServer.class.getResourceAsStream("keystore_server");
+        String clientKeyStorePass = "secret";
+
+        try {
+            sslReadyHttpAuthClient(clientKeyStore, clientKeyStorePass).execute(new HttpGet("https://localhost:" + server.getPort() + "/asd"));
+            fail("Request should have failed as server does not accept this client certificate");
+        } catch (Exception e) {
+            assertThat(e, isA(IOException.class));
```

Comment:
Is `org.apache.http.NoHttpResponseException` the right exception to be expected here?

Reply:
It is actually `SocketException` as it fails during socket connection. I changed to that instead.